smartsim/protoclient.py

```diff
@@ -0,0 +1,342 @@
+
+        value_array = numpy.array(self._timings.values(), dtype=float)
+        value_array = numpy.transpose(value_array)
+        for i in range(value_array.shape[0]):
+            print(" ".join(self._format_number(value) for value in value_array[i]))
+        if to_file:
+            numpy.save("timings.npy", value_array)
+            numpy.savetxt("timings.txt", value_array)
+
+    def run_model(self, model: t.Union[bytes, str], batch: torch.Tensor) -> t.Any:
+        """Execute a bach of inference requests with the supplied ML model.
+
+        :param model: The raw bytes or path to a pytorch model
+        :param batch: The tensor batch to perform inference on
+        :returns: The inference results
+        :raises ValueError: if the worker queue is not configured properly
+        in the environment variables
+        """
+        tensors = [batch.numpy()]
+        self.perf_timer.start_timings("batch_size", batch.shape[0])
+        built_tensor_desc = MessageHandler.build_tensor_descriptor(
+            "c", "float32", list(batch.shape)
+        )
+        self.perf_timer.measure_time("build_tensor_descriptor")
+        if isinstance(model, str):
+            model_arg = MessageHandler.build_model_key(model, self._backbone.descriptor)
+        else:
+            model_arg = MessageHandler.build_model(
+                model, "resnet-50", "1.0"
+            )  # type: ignore
+        request = MessageHandler.build_request(
+            reply_channel=self._from_worker_ch.descriptor,
+            model=model_arg,
+            inputs=[built_tensor_desc],
+            outputs=[],
+            output_descriptors=[],
+            custom_attributes=None,
+        )
+        self.perf_timer.measure_time("build_request")
+        request_bytes = MessageHandler.serialize_request(request)
+        self.perf_timer.measure_time("serialize_request")
+
+        if self._to_worker_fli is None:
+            raise ValueError("No worker queue available.")
+
+        # pylint: disable-next=protected-access
+        with self._to_worker_fli._channel.sendh(  # type: ignore
+            timeout=None,
+            stream_channel=self._to_worker_ch.channel,
+        ) as to_sendh:
+            to_sendh.send_bytes(request_bytes)
+            self.perf_timer.measure_time("send_request")
+            for tensor in tensors:
+                to_sendh.send_bytes(tensor.tobytes())  # TODO NOT FAST ENOUGH!!!
+                # to_sendh.send_bytes(bytes(tensor.data))
+        logger.info(f"Message size: {len(request_bytes)} bytes")
+
+        self.perf_timer.measure_time("send_tensors")
+        with self._from_worker_ch.channel.recvh(timeout=None) as from_recvh:
+            resp = from_recvh.recv_bytes(timeout=None)
+            self.perf_timer.measure_time("receive_response")
+            response = MessageHandler.deserialize_response(resp)
+            self.perf_timer.measure_time("deserialize_response")
+            # list of data blobs?
```

Comment:
Looks like this is a question! -> # list of data blobs?

Reply:
comment leftover from MVP removed